File: ruby/084.rb
# @param {Integer[]} heights
# @return {Integer}
def largest_rectangle_area(heights)
  result = 0

  past_heights = []
  heights.each do |height|
    if past_heights.length == 0 || past_heights[-1] <= height
      past_heights.push(height)
    else
      count = 0
      while past_heights.length > 0 && past_heights[-1] > height do
        count += 1
        h = past_heights.pop
        result = h * count if h * count > result
      end

      (count + 1).times do
        past_heights.push(height)
      end
    end
  end

  0.upto(past_heights.length - 1) do |i|
    value = past_heights[i] * (past_heights.length - i)
    result = value if value > result 
  end

  result
end

File: javascript/084.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function(heights) {
  let stack = [];
  let result = 0;

  heights.forEach(height => {
    if (stack.length === 0 || stack[stack.length - 1] <= height) {
      stack.push(height);
    } else {
      let count = 0;
      while (stack.length > 0 && stack[stack.length - 1] > height) {
        ++count;
        const h = stack.pop();
        if (h * count > result) result = h * count;
      }

      while (count >= 0) {
        stack.push(height);
        --count;
      }
    }
  });

  const length = stack.length;
  for(let i = 0; i < length; ++i) {
    const value = stack[i] * (length - i);
    if (value > result) result = value;
  }

  return result;
};